src/hotspot/share/classfile/systemDictionary.cpp

```diff
@@ -503,11 +503,12 @@ static void handle_parallel_super_load(Symbol* name,
                                        Handle protection_domain, TRAPS) {
 
   // superk is not used; resolve_super_or_fail is called for circularity check only.
-  Klass* superk = SystemDictionary::resolve_super_or_fail(name,
+  assert (!CDSConfig::is_dumping_static_archive(), "should not be parallel loading static archive");
+  Klass* superk = SystemDictionary::resolve_with_circularity_detection(name,
                                                           superclassname,
                                                           class_loader,
                                                           protection_domain,
-                                                          true,
+                                                          false,
```

Comment:
Please explain this change!

Reply:
Yes, this looks like an optimization, that we could do unconditionally if we want or not at all.  The comment mentions that we'll get a linkage error if we don't do it but I think it's talking about something else.

      // To support parallel loading: if class is done loading, just return the superclass
      // if the next_name matches class->super()->name() and if the class loaders match.
      // Otherwise, a LinkageError will be thrown later.
      if (klassk != nullptr && is_superclass &&
          ((quicksuperk = klassk->java_super()) != nullptr) &&
           ((quicksuperk->name() == next_name) &&
              (quicksuperk->class_loader() == class_loader()))) {
             return quicksuperk;
       } else {
        // Must check ClassCircularity before checking if superclass is already loaded.